### PR TITLE
[RDY] Treesitter set ranges

### PR DESCRIPTION
**Pull request by vigoux**
This PR adds a way to set the included ranges for a parser.

This is a required step towards language injection, which will later be done using lua only (hopefully...)

cc @tjdevries

### CODE REVIEW

**Comment by tjdevries** [comment]
I have no checked this over closely -- One thing that worries me is it looks like it depends directly on the buf and I'd like to make the implementation potentially not require a buf to do this (like for my parse string PR.)

Will this still allow for that later?

**Comment by vigoux** [comment]
> I have no checked this over closely -- One thing that worries me is it looks like it depends directly on the buf and I'd like to make the implementation potentially not require a buf to do this (like for my parse string PR.)
> 
> Will this still allow for that later?

Sounds quite hard to know... The buffer part is used only when trying to figure out the correct byte indexes (which is not really greatly done right now...). I think that we might want to have another function for that, or to provide the byte indexes within the `ranges` table.

File: src/nvim/lua/treesitter.c [hunk 1]
```diff
@@ -377,6 +385,101 @@ static int parser_edit(lua_State *L)
...
+    unsigned int stop_col = lua_tointeger(L, -1);
+    lua_pop(L, 1);  // [ parser, ranges, range ]
+
+    buf_T * buf = buflist_findnr(bufnr);
+
+    if (!buf) {
+      buf = curbuf;
+    }
+
+    // TODO(vigoux): For sure that's wrong, try to find a way to get the
```
Comment (tjdevries) [inline src/nvim/lua/treesitter.c:454]: Is this TODO still valid?

Comment (vigoux) [reply src/nvim/lua/treesitter.c:454]: Depends if we keep calculating the byte offsets here or if we move it somewhere else.

Comment (vigoux) [reply src/nvim/lua/treesitter.c:454]: I will me changing things to use nodes instead of raw ranges, this will be far more flexible with nodes

**Comment by vigoux** [comment]
It seems to work now, we use nodes to denote the start and the end of a range, and by that, it is not related to the source of the syntax tree anymore, and it should be usable with strings.
I did not tested it though.

File: src/nvim/lua/treesitter.c [hunk 2]
```diff
@@ -377,6 +385,86 @@ static int parser_edit(lua_State *L)
...
+    lua_rawgeti(L, 2, index + 1);  // [ parser, ranges, range ]
+
+    if (!lua_istable(L, -1)) {
+      xfree(ranges);
+      return luaL_error(
+          L,
+          "argument for parser:set_included_ranges() should be a table of tables.");
+    }
+
+    if (lua_objlen(L, -1) < 2) {
```
Comment (tjdevries) [inline src/nvim/lua/treesitter.c:422]: Probably `!=` rather than `<`

File: test/functional/lua/treesitter_spec.lua [hunk 3]
```diff
@@ -404,4 +404,57 @@ static int nlua_schedule(lua_State *const lstate)
...
+    local res = exec_lua([[
+    local root = parser:parse():root()
+    parser:set_included_ranges({{root:child(0), root:child(0)}})
+    parser.valid = false
+    return { parser:parse():root():range() }
+    ]])
+
+    eq({0, 0, 18, 1}, res)
+
+    -- Pick random samples
```
Comment (tjdevries) [inline test/functional/lua/treesitter_spec.lua:428]: Probably not actually "random", right?

File: test/functional/lua/treesitter_spec.lua [hunk 4]
```diff
@@ -404,4 +404,57 @@ static int nlua_schedule(lua_State *const lstate)
...
+    local res = exec_lua([[
+    parser = vim.treesitter.get_parser(0, "c")
+    return { parser:parse():root():range() }
+    ]])
+
+    eq({0, 0, 19, 0}, res)
+
+    local res = exec_lua([[
+    local root = parser:parse():root()
+    parser:set_included_ranges({{root:child(0), root:child(0)}})
```
Comment (tjdevries) [inline test/functional/lua/treesitter_spec.lua:421]: Can you give just a little info about this?

Also, can you split them into individual `it` tests, so you can say clearly what each is doing?

**Review by tjdevries — CHANGES_REQUESTED** [review]
Some small changes, but I think we're at least close to being good here.

File: src/nvim/lua/treesitter.c (hunk 2 above)
Comment (vigoux) [reply src/nvim/lua/treesitter.c:422]: Oh yeah of course... Ho can I miss that ...

File: test/functional/lua/treesitter_spec.lua (hunk 3 above)
Comment (vigoux) [reply test/functional/lua/treesitter_spec.lua:428]: Yes you are right

File: test/functional/lua/treesitter_spec.lua (hunk 4 above)
Comment (vigoux) [reply test/functional/lua/treesitter_spec.lua:423]: Yes, it should be done now !

Comment (teto) [reply test/functional/lua/treesitter_spec.lua:423]: @vigoux and I were discussing the API of `set_included_ranges` which is now called as an array of tuple `{ {node_start, node_end},  .., {node_start, node_end}` . I was triggered by the use of the same node here `root:child(0), root:child(0)}` and proposed to either pass a list of ranges like `set_included_ranges( {create_range(node_start,node_end),  create_range(node_start2,node_end2)} )` or just pass a list of node and use their ranges as done in https://tree-sitter.github.io/tree-sitter/using-parsers#multi-language-documents. @tjdevries what do you think ?

Comment (vigoux) [reply test/functional/lua/treesitter_spec.lua]: As discussed, we are going for a `table of nodes` solution, as it seems to be simpler on both implementation and usage. More complex ranges will in any case be easy to generate by iterating over each intermediate node between two nodes.
Something else to note is that this is used mainly for language injection, and most of the parsers, have clearly identified nodes that needs language injection on them.

**Comment by tjdevries** [comment]
LGTM, thanks for the review @teto and nice change to the nodes method. Very nice improvement from our first iteration!

**Comment by vigoux** [comment]
Yeah ! That version is much simpler both to use and to implement ! And that's great !
Next step is treesitter highlighting !

**Pull request merged**